fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1056,329 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+	std::unordered_map<TenantName, TenantMapEntry> mgmtClusterTenantMap;
+	std::unordered_map<uint64_t, TenantName> mgmtClusterTenantIdIndex;
+	std::unordered_set<TenantNameRef> mgmtTenantSet;
+
+	RestoreClusterImpl(Reference<DB> managementDb,
+	                   ClusterName clusterName,
+	                   ClusterConnectionString connectionString,
+	                   DataClusterEntry clusterEntry,
+	                   AddNewTenants addNewTenants,
+	                   RemoveMissingTenants removeMissingTenants)
+	  : ctx(managementDb, clusterName), clusterName(clusterName), connectionString(connectionString),
+	    clusterEntry(clusterEntry), addNewTenants(addNewTenants), removeMissingTenants(removeMissingTenants) {}
+
+	ACTOR static Future<bool> markClusterAsRestoring(RestoreClusterImpl* self,
+	                                                 Reference<typename DB::TransactionT> tr) {
+		if (self->ctx.dataClusterMetadata.get().entry.clusterState != DataClusterState::RESTORING) {
+			DataClusterEntry updatedEntry = self->ctx.dataClusterMetadata.get().entry;
+			updatedEntry.clusterState = DataClusterState::RESTORING;
+
+			updateClusterMetadata(tr,
+			                      self->ctx.clusterName.get(),
+			                      self->ctx.dataClusterMetadata.get(),
+			                      Optional<ClusterConnectionString>(self->connectionString),
+			                      updatedEntry);
+		}
+
+		TraceEvent("MarkedDataClusterRestoring").detail("Name", self->ctx.clusterName.get());
+
+		return true;
+	}
+
+	ACTOR static Future<Void> markClusterAsReady(RestoreClusterImpl* self, Reference<typename DB::TransactionT> tr) {
+		if (self->ctx.dataClusterMetadata.get().entry.clusterState == DataClusterState::RESTORING) {
+			DataClusterEntry updatedEntry = self->ctx.dataClusterMetadata.get().entry;
+			updatedEntry.clusterState = DataClusterState::READY;
+
+			updateClusterMetadata(tr,
+			                      self->ctx.clusterName.get(),
+			                      self->ctx.dataClusterMetadata.get(),
+			                      Optional<ClusterConnectionString>(self->connectionString),
+			                      updatedEntry);
+		}
+
+		TraceEvent("MarkedDataClusterReady")
+		    .detail("Name", self->ctx.clusterName.get())
+		    .detail("Version", tr->getCommittedVersion());
+
+		return Void();
+	}
+
+	ACTOR static Future<Void> markManagementTenantAsError(Reference<typename DB::TransactionT> tr,
+	                                                      TenantName tenantName) {
+		state Optional<TenantMapEntry> tenantEntry = wait(tryGetTenantTransaction(tr, tenantName));
+
+		if (!tenantEntry.present()) {
+			return Void();
+		}
+
+		tenantEntry.get().tenantState = TenantState::ERROR;
+		ManagementClusterMetadata::tenantMetadata().tenantMap.set(tr, tenantName, tenantEntry.get());
+		return Void();
+	}
+
+	ACTOR static Future<bool> getTenantsFromDataCluster(RestoreClusterImpl* self, Reference<ITransaction> tr) {
+		TenantNameRef begin = self->clusterName;
```

Comment:
I think here you can start on the empty tenant:

```suggestion
		TenantNameRef begin = ""_sr;
```